src/java.base/share/classes/jdk/internal/lang/CaseFolding.java.template

```diff
@@ -0,0 +1,209 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.lang;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static java.util.Map.entry;
+
+/**
+ * Utility class that handles Unicode case folding properties defined in
+ * CasingFolding.txt, including 1:M full case folding.
+ */
+public final class CaseFolding {
+
+    private CaseFolding()  {}
+
+   /**
+    * Tests whether the specified code point has a folding mapping entry defined.
+    *
+    * @param  cp
+    *         the Unicode code point to test
+    * @return {@code true} if the given code point has a case folding mapping entry
+    *         defined in (@code caseFoldingMap}, {@code false} otherwise
+    */
+    public static boolean isDefined(int cp) {
+         return getDefined(cp) != -1;
+     }
+
+   /**
+    * Returns the case-folded form of the specified code point according
+    * to the Unicode case folding mappings.
+    * <p>
+    * If the code point has no case folding mapping defined, this method returns
+    * the original code point.
+    *
+    * Possible combinations of the returning case-folding form as a long value
+    *
+    *  +---+---------+--------+---------+--------+--------+
+    *  | 1:1 mapping |  0000  |   0000  |  000x  |  xxxx  |  0041 => 0061 or 1E921 => 1E943
+    *  +---+---------+--------+---------+--------+--------+
+    *  | 1:2 mapping |  0002  |   0000  |  xxxx  |  xxxx  |  FB02 => 0066 006C
+    *  +---+---------+--------+---------+--------+--------+
+    *  | 1:3 mapping |  0003  |   xxxx  |  xxxx  |  xxxx  |  FB03 => 0066 0066 0069
+    *  +---+---------+--------+---------+--------+--------+
+    *
+    * @param  cp
+    *         the Unicode code point to fold
+    * @return a long value representing the case-folded form of the input
+    *         code point, encoded as TBD
+    */
+    public static long fold(int cp) {
+        var fold = getDefined(cp);
+        return fold == -1 ? cp : fold;
+    }
+
+    public static boolean isSingleCodePoint(long fold) {
+        return (fold >> 48) == 0;
+    }
+
+   /**
+    * Returns an expansion set to "close" a given regex Unicode character class range for case-sensitive
+    * matching, according to the
+    * <a href="https://www.unicode.org/reports/tr18/#Simple_Loose_Matches">Simple Loose Matches</a>
+    * rule defined in Unicode Technical Standard #18: Unicode Regular Expressions.
+    * <p>
+    * To conform with Level 1 of UTS #18, specifically RL1.5: Simple Loose Matches, simple case folding must
+    * be applied to literals and (optionally) to character classes. When applied to character classes, each
+    * character class is expected to be closed under simple case folding. See the standard for the
+    * detailed explanation and example of "closed".
```

Comment:
A helpful explanation of "close" would be useful here to avoid having to chase down and understand the standard.